File: bucketSort.java


import java.util.Scanner;
import java.io.*;

public class bucketSort{
	
	private int min, max, lastIndex;
	private int[] BucketAry;
	
	public bucketSort(int min, int max){
		this.min = min;
		this.max = max;
		lastIndex = (max-min);
		BucketAry = new int[lastIndex+1];
		for(int i=0; i<BucketAry.length; i++) { BucketAry[i] = 0; }
	}
	
	public void getIndex(int data){
		BucketAry[data-min]++;
	}
	
	public void printSortData(PrintWriter x){
		for(int i=0; i<BucketAry.length; i++){
			for(int j = BucketAry[i]; j!=0; j--){
				x.println(min+i);
				System.out.println(min+i);
			}
		}
		
		x.close();
	}
	
	public static int[] findMinMax(Scanner file){
		
		/* We initialize myArray with MAX_VALUE and MIN_VALUE 
		 * so we can do an initial test for the first input value from the txt file.
		 */
		
		int[] myArray = new int[2];
		myArray[0] = Integer.MAX_VALUE;			
		myArray[1] = Integer.MIN_VALUE;			
		int num;
		
		while(file.hasNext()){
			if(file.hasNextInt()){
				num = file.nextInt();
					if(num < 0){
						System.out.println("A negative value was read. Program will exit now.");
						System.exit(0);
					}
					if(num < myArray[0]){ myArray[0] = num; }
					if(num > myArray[1]){ myArray[1] = num; }
			}
		}
		return myArray;
	}
	
	public static void main(String[] args) throws FileNotFoundException{
		
		int number;
		int[] retArr; 
		Scanner readFile;
		File inFile;
		
		if(args.length > 0){
			inFile = new File(args[0]);
			readFile = new Scanner(inFile);
		}else{
			inFile = new File("Bucket_Sort_Data.txt");
			readFile = new Scanner(inFile);
		}
		
		
		retArr = bucketSort.findMinMax(readFile);
		readFile.close();
		System.out.println(retArr[0] + " " + retArr[1]);
		bucketSort sort = new bucketSort(retArr[0], retArr[1]);
		
		readFile = new Scanner(inFile);
		while(readFile.hasNext()){
			if(readFile.hasNextInt()){
				number = readFile.nextInt();
				sort.getIndex(number);
			}
		}
		
		readFile.close();
		
		if(args.length > 0){
			PrintWriter writer = new PrintWriter(args[1]);
			sort.printSortData(writer);
			writer.close();
		}else{
			PrintWriter writer = new PrintWriter("Sorted_Data.txt");
			sort.printSortData(writer);
			writer.close();
		}
	}
}
	
	
	

File: bucketSort.cpp
#include <iostream>
#include <fstream>
#include <climits>
#include <cstdlib>
using namespace std;

class bucketSort {
    int max, min, lastIndex;
    int *BucketAry;
    public:
        bucketSort(int x, int y) {
            min = x;
            max = y;
            lastIndex = (max-min);
            BucketAry = new int[lastIndex+1];
            for(int i = 0; i<lastIndex; i++){
                BucketAry[i] = 0;
            }
        }
        static int *findMinMax(fstream &inFile){
            int y=INT_MIN, x=INT_MAX, number;
            static int numAry[2];

            while(inFile >> number){
                if(number<0) {
                    cout << "Error: A negative value was read!" << endl;
                    exit(0);
                }
                if(number<x){
                    numAry[0] = number;
                    x = number;
                }
                if(number>y){
                    numAry[1] = number;
                    y = number;
                }
            }
                return numAry;
        }
        int getIndex(int data){
            BucketAry[data-min]++;
            return BucketAry[data-min];
        }
        void printSortData(fstream *outFile){
            for(int i = 0; i<=lastIndex; i++){
                for(int j = BucketAry[i]; j!= 0; j--){
                    *outFile << (min+i) << endl;
                }
            }
            outFile->close();
        }
};







int main(int argc, char* argv[])
{
    int number;
    int *numAry;
    fstream myFile;
    myFile.open(argv[1]);
    if(!myFile.is_open()){
        cout << "File did not open!" << endl;
        exit(0);
    }
    numAry = bucketSort::findMinMax(myFile);
    myFile.close();

    myFile.open(argv[1]);
    bucketSort *b = new bucketSort(numAry[0],numAry[1]);
    while(myFile >> number){
        b->getIndex(number);
    }

    myFile.close();
    myFile.open(argv[2]);
    b->printSortData(&myFile);
    myFile.close();
    delete b;

    return 0;

}

